File: src/main/java/com/vandendaelen/nicephore/enums/OperatingSystems.java
package com.vandendaelen.nicephore.enums;

import com.vandendaelen.nicephore.clipboard.ClipboardManager;
import com.vandendaelen.nicephore.clipboard.impl.MacOSClipboardManagerImpl;
import com.vandendaelen.nicephore.clipboard.impl.WindowsClipboardManagerImpl;
import net.minecraft.Util;

public enum OperatingSystems {
    WINDOWS(WindowsClipboardManagerImpl.getInstance()),
    LINUX(null),
    MAC(MacOSClipboardManagerImpl.getInstance()),
    SOLARIS(null);

    private final ClipboardManager manager;
    private static OperatingSystems OS;

    OperatingSystems(ClipboardManager instance) {
        this.manager = instance;
    }

    public static OperatingSystems getOS() {
        return switch (Util.getPlatform()) {
            case WINDOWS -> WINDOWS;
            case LINUX -> LINUX;
            case OSX -> MAC;
            case SOLARIS -> SOLARIS;
            default -> throw new IllegalStateException("Unexpected value: " + Util.getPlatform());
        };
    }

    public ClipboardManager getManager() {
        return manager;
    }
}

File: src/main/kotlin/com/vandendaelen/nicephore/enums/OperatingSystems.kt
package com.vandendaelen.nicephore.enums

import com.vandendaelen.nicephore.clipboard.ClipboardManager
import com.vandendaelen.nicephore.clipboard.impl.MacOSClipboardManagerImpl
import com.vandendaelen.nicephore.clipboard.impl.WindowsClipboardManagerImpl
import net.minecraft.Util

enum class OperatingSystems(val manager: ClipboardManager?) {
    WINDOWS(WindowsClipboardManagerImpl.instance),
    LINUX(null),
    MAC(MacOSClipboardManagerImpl.instance),
    SOLARIS(null);

    companion object {
        fun getOS(): OperatingSystems = when (Util.getPlatform()) {
            Util.OS.WINDOWS -> WINDOWS
            Util.OS.LINUX -> LINUX
            Util.OS.OSX -> MAC
            Util.OS.SOLARIS -> SOLARIS
            else -> throw IllegalStateException("Unexpected value: ${Util.getPlatform()}")
        }
    }
}
